store/sqlstore/post_store.go

```diff
@@ -2364,7 +2365,40 @@ func (s *SqlPostStore) cleanupThreads(postId, rootId string, permanent bool) err
 		return nil
 	}
 	if rootId != "" {
-		_, err := s.GetMaster().Exec(`UPDATE Threads SET ReplyCount = ReplyCount - 1 WHERE PostId = :Id AND ReplyCount > 0`, map[string]interface{}{"Id": rootId})
+		queryString, args, _ := s.getQueryBuilder().
+			Select("COUNT(Id)").
+			From("Posts").
+			Where(sq.And{
+				sq.Eq{"RootId": rootId},
+				sq.Eq{"UserId": userId},
+				sq.Eq{"DeleteAt": 0},
+			}).
+			ToSql()
+
+		count, err := s.GetReplica().SelectInt(queryString, args...)
+
+		if err != nil {
+			mlog.Warn("Error counting user's posts in thread", mlog.Err(err))
+		}
+
+		queryString, args, _ = s.getQueryBuilder().
+			Select("Participants").
+			From("Threads").
+			Where(sq.Eq{"PostId": rootId}).
+			ToSql()
+
+		var participants model.StringArray
+		err = s.GetReplica().SelectOne(&participants, queryString, args...)
+
+		if err != nil {
+			mlog.Warn("Error getting thread participants", mlog.Err(err))
+		}
+
+		if count == 0 && participants.Contains(userId) {
+			participants = participants.Remove(userId)
+		}
+
+		_, err = s.GetMaster().Exec(`UPDATE Threads SET ReplyCount = ReplyCount - 1, Participants = :Participants WHERE PostId = :Id AND ReplyCount > 0`, map[string]interface{}{"Id": rootId, "Participants": participants})
```

Comment:
A couple of things:

- We should probably aim at fetching (and setting of course) participants only when we actually have to update the list (count is zero). I think it should be possible using the squirrel query builder without causing too much code duplication.
- Since we made the `Threads.Participants` column to be of native JSON type we may want to explore a way to do the search and update on the DB side. This can be deferred to a later PR but something to keep in mind. /cc @agnivade

Reply:
> Since we made the `Threads.Participants` column to be of native JSON type we may want to explore a way to do the search and update on the DB side. This can be deferred to a later PR but something to keep in mind. /cc @agnivade

If you ever get to that please ping me, I would like to see how it's going to be implemented :)

Reply:
> We should probably aim at fetching (and setting of course) participants only when we actually have to update the list (count is zero). I think it should be possible using the squirrel query builder without causing too much code duplication.

Just a question, could the following be reduced in one line?

```go

			queryString, args, _ = s.getQueryBuilder().
				Select("Participants").
				From("Threads").
				Where(sq.Eq{"PostId": rootId}).
				ToSql()

			var participants model.StringArray
			err = s.GetReplica().SelectOne(&participants, queryString, args...)
```

Reply:
I don't think we ever made use of `RunWith()` but I guess it should work just fine.

Reply:
@koox00 - the code is already here:

https://github.com/mattermost/mattermost-server/blob/68517a77195664c15ee24e7b12677b0f66a94d5c/store/sqlstore/thread_store.go#L685-L703


We just need to remove from the array and invert the membership check. We added native JSON support to explicitly avoid double querying like this. I have been stuck with some other tasks and therefore have not been able to demo these new features.

But I'd love to see us use these features. Would you be open to creating an HW ticket to convert this to native JSON queries? Seems like an interesting ticket to have.

Reply:
Yep sure, added a ticket as a HW candidate https://mattermost.atlassian.net/browse/MM-39157